File: client/client-multi/client-objective-c/Kaa/channel/impl/transports/DefaultUserTransport.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "DefaultUserTransport.h"
#import "EndpointRegistrationProcessor.h"
#import "EndpointAccessToken.h"
#import "KaaLogging.h"

#define TAG @"DefaultUserTransport >>>"

@interface DefaultUserTransport ()

@property (nonatomic, strong) id<EndpointRegistrationProcessor> processor;
@property (nonatomic, strong) NSMutableDictionary *attachedEndpoints;

@end

@implementation DefaultUserTransport

- (instancetype)init {
    self = [super init];
    if (self) {
        self.attachedEndpoints = [NSMutableDictionary dictionary];
    }
    return self;
}

- (UserSyncRequest *)createUserRequest {
    if (!self.processor) {
        DDLogWarn(@"%@ Unable to create UserSyncRequest - no EndpointRegistrationProcessor specified!", TAG);
        return nil;
    }
    
    NSDictionary *attachEndpointRequests = [self.processor getAttachEndpointRequests];
    NSMutableArray *attachEPRequestList = [NSMutableArray array];
    for (NSNumber *key in attachEndpointRequests.allKeys) {
        EndpointAttachRequest *attachRequest = [[EndpointAttachRequest alloc] init];
        attachRequest.requestId = [key intValue];
        attachRequest.endpointAccessToken = [attachEndpointRequests[key] token];
        [attachEPRequestList addObject:attachRequest];
    }
    
    NSDictionary *detachEndpointRequests = [self.processor getDetachEndpointRequests];
    NSMutableArray *detachEPRequestList = [NSMutableArray array];
    for (NSNumber *key in detachEndpointRequests.allKeys) {
        EndpointDetachRequest *detachRequest = [[EndpointDetachRequest alloc] init];
        detachRequest.requestId = [key intValue];
        detachRequest.endpointKeyHash = [detachEndpointRequests[key] keyHash];
        [detachEPRequestList addObject:detachRequest];
    }
    
    UserSyncRequest *request = [[UserSyncRequest alloc] init];
    if ([self.processor getUserAttachRequest]) {
        request.userAttachRequest = [KAAUnion unionWithBranch:KAA_UNION_USER_ATTACH_REQUEST_OR_NULL_BRANCH_0
                                                         data:[self.processor getUserAttachRequest]];
    }
    request.endpointAttachRequests = [KAAUnion unionWithBranch:KAA_UNION_ARRAY_ENDPOINT_ATTACH_REQUEST_OR_NULL_BRANCH_0
                                                          data:attachEPRequestList];
    request.endpointDetachRequests = [KAAUnion unionWithBranch:KAA_UNION_ARRAY_ENDPOINT_DETACH_REQUEST_OR_NULL_BRANCH_0
                                                          data:detachEPRequestList];
    return request;
}

- (void)onUserResponse:(UserSyncResponse *)response {
    if (!self.processor) {
        return;
    }
    BOOL hasChanges = NO;
    if (response.endpointAttachResponses
        && response.endpointAttachResponses.branch == KAA_UNION_ARRAY_ENDPOINT_ATTACH_RESPONSE_OR_NULL_BRANCH_0) {
        NSDictionary *attachEndpointRequests = [self.processor getAttachEndpointRequests];
        NSArray *attachResponses = response.endpointAttachResponses.data;
        for (EndpointAttachResponse *attached in attachResponses) {
            EndpointAccessToken *attachedToken = attachEndpointRequests[@(attached.requestId)];
            if (attached.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS && attachedToken) {
                DDLogInfo(@"%@ Token: %@", TAG, attachedToken);
                if (attached.endpointKeyHash.branch == KAA_UNION_STRING_OR_NULL_BRANCH_0) {
                    EndpointKeyHash *keyHash = [[EndpointKeyHash alloc] initWithKeyHash:attached.endpointKeyHash.data];
                    self.attachedEndpoints[attachedToken] = keyHash;
                    hasChanges = YES;
                } else {
                    DDLogError(@"%@ No endpointKeyHash for request id: %i", TAG, attached.requestId);
                }
            } else {
                DDLogError(@"%@ Failed to attach endpoint with token: %@. Request id: %i, result: %i",
                           TAG, attachedToken, attached.requestId, attached.result);
            }
        }
    }
    
    if (response.endpointDetachResponses
        && response.endpointDetachResponses.branch == KAA_UNION_ARRAY_ENDPOINT_DETACH_RESPONSE_OR_NULL_BRANCH_0) {
        NSDictionary *detachEndpointRequests = [self.processor getDetachEndpointRequests];
        NSArray *detachResponses = response.endpointDetachResponses.data;
        for (EndpointDetachResponse *detached in detachResponses) {
            EndpointKeyHash *detachedEndpointKeyHash = detachEndpointRequests[@(detached.requestId)];
            if (detached.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS && detachedEndpointKeyHash) {
                for (EndpointAccessToken *key in self.attachedEndpoints.allKeys) {
                    EndpointKeyHash *value = self.attachedEndpoints[key];
                    if ([value isEqual:detachedEndpointKeyHash]) {
                        [self.attachedEndpoints removeObjectForKey:key];
                        if (!hasChanges) {
                            hasChanges = YES;
                        }
                    }
                }
            } else {
                DDLogError(@"%@ Failed to detach endpoint with key hash: %@. Request id: %i, result: %i",
                           TAG, detachedEndpointKeyHash, detached.requestId, detached.result);
            }
        }
    }
    
    if (hasChanges && self.clientState) {
        [self.clientState setAttachedEndpoints:self.attachedEndpoints];
    }
    
    [self.processor onUpdateWithAttachResponses:response.endpointAttachResponses.data
                                detachResponses:response.endpointDetachResponses.data
                                   userResponse:response.userAttachResponse.data
                         userAttachNotification:response.userAttachNotification.data
                         userDetachNotification:response.userDetachNotification.data];
    
    DDLogInfo(@"%@ Processed user response", TAG);
}

- (void)setEndpointRegistrationProcessor:(id<EndpointRegistrationProcessor>)processor {
    self.processor = processor;
}

- (TransportType)getTransportType {
    return TRANSPORT_TYPE_USER;
}

@end

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/channel/impl/transports/DefaultUserTransport.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.channel.impl.transports;

import org.kaaproject.kaa.client.channel.UserTransport;
import org.kaaproject.kaa.client.event.EndpointAccessToken;
import org.kaaproject.kaa.client.event.EndpointKeyHash;
import org.kaaproject.kaa.client.event.registration.EndpointRegistrationProcessor;
import org.kaaproject.kaa.common.TransportType;
import org.kaaproject.kaa.common.endpoint.gen.EndpointAttachRequest;
import org.kaaproject.kaa.common.endpoint.gen.EndpointAttachResponse;
import org.kaaproject.kaa.common.endpoint.gen.EndpointDetachRequest;
import org.kaaproject.kaa.common.endpoint.gen.EndpointDetachResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.kaaproject.kaa.common.endpoint.gen.UserSyncRequest;
import org.kaaproject.kaa.common.endpoint.gen.UserSyncResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class DefaultUserTransport extends AbstractKaaTransport implements
    UserTransport {

  private static final Logger LOG = LoggerFactory.getLogger(DefaultUserTransport.class);

  private EndpointRegistrationProcessor processor;
  private Map<EndpointAccessToken, EndpointKeyHash> attachedEndpoints = new HashMap<>();

  @Override
  public UserSyncRequest createUserRequest() {
    if (processor != null) {
      UserSyncRequest request = new UserSyncRequest();

      Map<Integer, EndpointAccessToken> attachEndpointRequests =
              processor.getAttachEndpointRequests();
      List<EndpointAttachRequest> attachEpRequestList = new LinkedList<EndpointAttachRequest>();
      for (Map.Entry<Integer, EndpointAccessToken> attachEpRequest : attachEndpointRequests
              .entrySet()) {
        attachEpRequestList.add(new EndpointAttachRequest(attachEpRequest.getKey(),
                attachEpRequest.getValue().getToken()));
      }

      Map<Integer, EndpointKeyHash> detachEndpointRequests = processor.getDetachEndpointRequests();
      List<EndpointDetachRequest> detachEpRequestList = new LinkedList<>();
      for (Map.Entry<Integer, EndpointKeyHash> detachEpRequest : detachEndpointRequests
              .entrySet()) {
        detachEpRequestList.add(new EndpointDetachRequest(detachEpRequest.getKey(),
                detachEpRequest.getValue().getKeyHash()));
      }

      request.setEndpointAttachRequests(attachEpRequestList);
      request.setEndpointDetachRequests(detachEpRequestList);
      request.setUserAttachRequest(processor.getUserAttachRequest());
      return request;
    }
    return null;
  }

  @Override
  public void onUserResponse(UserSyncResponse response) throws IOException {
    if (processor != null) {
      boolean hasChanges = false;
      if (clientState != null) {
        attachedEndpoints = clientState.getAttachedEndpointsList();
      }
      Map<Integer, EndpointAccessToken> attachEndpointRequests =
              processor.getAttachEndpointRequests();
      if (response.getEndpointAttachResponses() != null
              && !response.getEndpointAttachResponses().isEmpty()) {
        for (EndpointAttachResponse attached : response.getEndpointAttachResponses()) {
          EndpointAccessToken attachedToken = attachEndpointRequests.remove(
                  attached.getRequestId());
          if (attached.getResult() == SyncResponseResultType.SUCCESS) {
            if (attachedToken != null) {
              LOG.info("Token {}", attachedToken);
              attachedEndpoints.put(attachedToken,
                      new EndpointKeyHash(attached.getEndpointKeyHash()));
              hasChanges = true;
            } else {
              LOG.warn("Endpoint {} is already attached!", attached.getEndpointKeyHash());
            }
          } else {
            LOG.error("Failed to attach endpoint {}. Attach endpoint request id: {}",
                    attached.getEndpointKeyHash(), attached.getRequestId());
          }
        }
      }
      Map<Integer, EndpointKeyHash> detachEndpointRequests = processor.getDetachEndpointRequests();
      if (response.getEndpointDetachResponses() != null
              && !response.getEndpointDetachResponses().isEmpty()) {
        for (EndpointDetachResponse detached : response.getEndpointDetachResponses()) {
          EndpointKeyHash detachedEndpointKeyHash = detachEndpointRequests.remove(
                  detached.getRequestId());
          if (detached.getResult() == SyncResponseResultType.SUCCESS) {
            if (detachedEndpointKeyHash != null) {
              for (Map.Entry<EndpointAccessToken, EndpointKeyHash> entry : attachedEndpoints
                      .entrySet()) {
                if (detachedEndpointKeyHash.equals(entry.getValue())) {
                  EndpointKeyHash removed = attachedEndpoints.remove(entry.getKey());
                  if (!hasChanges) {
                    hasChanges = (removed != null);
                  }
                  break;
                }
              }
            }
          } else {
            LOG.error("Failed to detach endpoint. Detach endpoint request id: {}",
                    detached.getRequestId());
          }
        }
      }

      if (hasChanges && clientState != null) {
        clientState.setAttachedEndpointsList(attachedEndpoints);
      }
      processor.onUpdate(response.getEndpointAttachResponses(),
              response.getEndpointDetachResponses(),
              response.getUserAttachResponse(),
              response.getUserAttachNotification(),
              response.getUserDetachNotification());
      LOG.info("Processed user response");
    }
  }

  @Override
  public void setEndpointRegistrationProcessor(EndpointRegistrationProcessor manager) {
    this.processor = manager;
  }

  @Override
  protected TransportType getTransportType() {
    return TransportType.USER;
  }

}
